### ISSUE TITLE
Scope of team

### ISSUE THREAD

**Issue by MylesBorins** [nodejs/modules#17]
If we plan to charter, part of the process will be deciding on the scope of the team, and what we will be explicitly asking the TSC to take responsibility for.

While the original scope of this group was ESM, it is very likely that part of (or potentially all) the CJS implementation may be required to be in the scope of this group.

Let's use this thread to brainstorm

**Comment by devsnek** [comment]
i would like to see this group handling everything related to how modules of the esm and cjs (and whatever else?) get handled in node, as they are definitely coupled in the ecosystem

**Comment by bmeck** [comment]
I have a few topics that I would like to stick up for discussion that need to be cross module system:

* content type conflicts, namely:
    * file extension conflict resolution
    * URL affects for things like `blob:` / `text:` and the affects on being in conflict with loader hook results
* loader hooks (per https://github.com/nodejs/modules/issues/6)
* integrity checks as a future proofed need

**Comment by zackschuster** [comment]
building on from @bmeck, from what i've seen so far, here are the major asks:

- transparent interop between ESM + CJS w/o a build step
- transparent interop between Node/Browsers when running ESM
- [allow consumers to intercept/override `require` ](https://github.com/nodejs/modules/issues/6)
- fallback mechanism for legacy modules to be converted to ESM without losing access to scoped variables

**Comment by MylesBorins** [comment]
So I think that the high level take away at this point is that our scope is going to be both ESM + CJS. The group may in turn take responsibility for all of require, the module specifier resolution algorithm, etc.

I do have slight concerns about how quickly this could expand and move away from the intent and timely work. Would it perhaps make sense to scope to ESM + interoperability with CJS?

**Comment by zackschuster** [comment]
given that merely _accepting a new file extension_ has set everyone back by god-knows-how-long, i'm inclined to say yes, we should define an extraordinarily limited MVP or the group may never ship.

**Comment by WebReflection** [comment]
> not saying `.mjs` is bad

I think you can say that, we're here because nobody liked that. Not bad as workaround, bad as solution for the community.

> we should define an extraordinarily limited MVP or the group may never ship

amen.

**Comment by WebReflection** [comment]
> Would it perhaps make sense to scope to ESM + interoperability with CJS

It's the only way forward if we want to even think about shipping in v10.

**Comment by bmeck** [comment]
> I think you can say that, we're here because nobody liked that.

I think we should agree to disagree on a variety of things.

**Comment by zackschuster** [comment]
@WebReflection i'd agree, `.mjs` has a lot of problems (my favorite is breaking file previews) but i don't want to discount people who prefer that approach for their own reasons. 

beyond that, being "realistic" i'm guessing it'll still be an uphill battle for an alternative proposal to succeed, if only because "`.mjs` works" and therefore there's likely a massive amount of pressure from interested parties in seeing that move forward regardless -- which would also mean there's a very real possibility that, even if the group ships, steering may decline the product.

**Comment by WebReflection** [comment]
it's demonstrated, and easy to demonstrate, that `.mjs` doesn't work. It's a convenient workaround, not a migration pattern, neither a solution. Nobody writes extensions in Node, which is first AFAIK agreed point of this team: make ESM that works in browsers work out of  the box in NodeJS.

Pure ESM in browsers is `import mod from "./file.js"` and it works already, is not something that needs to ship, differently from node, rather something already shipped, deployed, and working without any issue.

The hooks on web are on loading, like `type=module`, not on mime-type or file extensions, so there's no need to go down that ugly file extension road what won't seamlessly work on the Web where ESM is already almost everywhere without any flag.

So, _ESM + interoperability with CJS_ seems to be right team scope, and _we should define an extraordinarily limited MVP_ the only way this group of people could make sense.

**Comment by bmeck** [comment]
@WebReflection we completely disagree on these topics and conclusions.

**Comment by WebReflection** [comment]
there's nothing to disagree, I've presented always facts, never opinions. When you'll point at me at a well established NodeJS project that uses file extensions in CommonJS or transpiled fake ESM then you might have some counter fact to talk about. until that, I use already unpkg.com CDN to deliver pure ESM and being unpkg.com a mirror of npm my modules work already in ESM capable environments (SpiderMonkey, JSC) and are also transpiled to CommonJS for Node.

Basically I already experience the goal of this team and I can say it works already.

Same goes for everyone else that used `@std/esm` without even needing full file resolutions on ESM.

**Comment by MylesBorins** [comment]
@WebReflection userland and platform implementations have extremely different requirement. Your statement regarding what can already be accomplished with tooling is proof that a vibrant ecosystem can come up with great solutions

Here's a scope I would like to make and challenge everyone to abide to

Can we please scope our conversation to explicit proposals rather than philosophy. We continue to get stuck in the weeds of implementation details and focusing on parts of the user experience that we find to not be ergonomic.

For example I would not like to ship transparent interop. That being said I think I will have a better chance of influencing the decision of what we ship by suggesting concrete changes (import.meta.require) and [suggesting compromises](https://github.com/nodejs/node/pull/18392#issuecomment-3634441400) that can work towards my goal.

This group was spun up to come up with concrete proposals that find us a path forward. If we kick this off with infighting and endless debate around implementation we will get nowhere.

**Comment by WebReflection** [comment]
FWIW I'm already pro `import.meta.require` (even if I've discovered you cannot actually pollute `import.meta` in current WebKit/Chrome browsers) and fully happy with a `node --type=module` equivalent.

To me that is the best deal that works with pure ESM that already works in browsers and works with CJS modules.

Ask me if I'd like that any time you want, my answer will be always yes.

**Comment by WebReflection** [comment]
To strength the feasibility:

`js52 -m PATH` or `js52 --module=PATH` already works with `.js` files in every Linux shell.

`jsc -m PATH` or `js52 --module-file=PATH` already works with `.js` files in every Linux shell.

[NexusJS](https://github.com/voodooattack/nexusjs) which is a competitor (big selling point?) already works with everything as expected (and `.js` **is** the expected for both ESM or CJS).

Browsers `<script type=module>` already work with ESM too. `text/javascript` or `application/javascript` are valid ESM.

We miss interoperability with CommonJS. While all the mentioned environments already have an ESM that brings CJS in, NodeJS could pave the road for `import.meta.require` which as suggested is static analysis friendly.

With only these two as proposal and goals we restrict our scope and we can ship for v10.

With anything bigger than 2 or 3 bullet points and 3 months I don't think there's anything we can do for v10 so we can take all the time we want for v12 (undesirable, imo)

**Comment by MylesBorins** [comment]
@WebReflection the deadline for v10 is closer to 9 - 27 months. We can remove the flag any time and land as semver-minor. Only semver major changes against core will have that cut off (and it actually is more like 2 months)

Can we please keep this issue on track though... focusing on scope, not implementation

**Comment by ljharb** [comment]
> For example I would not like to ship transparent interop.

For me, transparent interop is a critical requirement.

**Comment by evanplaice** [comment]
Can we nail down some initial milestones? Much of the work that needs to be done requires additional research, exploration, and discovery. Here's a draft that attempts to provide structure to the issues discussed in @ceejbot's proposal and in the issue comments so far.

### Milestone 0 - target [current] -  CommonJS in Node

**Goal: Ensure backward compatibility**

- leave all existing built-ins as is (don't modify existing globals in Node)

**Details**

It's unrealistic (at this time) to change existing built-ins without breaking the current node module ecosystem. It's not productive to even beginning to discuss deprecating CJS until ESM has been implemented and used extensively in the Node ecosystem.

This could be left unsaid but it's important to establish intent. This group doesn't exist to obsolete CJS without consideration of the greater Node community.

### Milestone 1 - target [v10.0] - ESM Support in Node

**Goal: ESM modules work in Node without plugins, experimental flags, loaders**

- [ ] ESM can be automatically detected and loaded
- [ ] CJS can be automatically detected and loaded
- [ ] mode:esm can be explicitly set by a CLI flag
- [ ] mode:cjs can be explicity set by a CLI flag
- [ ] mode:esm can be explicitly defined in package.json
- [ ] mode:cjs (default) can be explicitly defined in package.json
- [ ] Node throws error if CJS is imported while mode:esm
- [ ] Node throws an error if ESM is imported while mode:cjs

**Details:**

This is only the first step. It **does not** include ESM<->CJS interop. It focuses on providing ESM support in Node. Any attempt to mix ESM with CJS or vice-versa will throw an error.

10.0 is a notable release number and provides an opportunity to mark a big transition point. For example, OS9 vs OSX. It can be used to mark a transition. For example, there is a huge body of documented questions on StackOverflow addressing existing CJS workflows. If the introduction of ESM happens with 10.0, it should be easy to tag all Node questions on StackOverflow with pre-10 vs post-10 context tags.

This will provide an opportunity for early-adopters to experiment with ESM in Node.js. It may also provide an opportunity for eager devs to produce alternative workarounds that address dealing with ESM<->CJS interop.

### Milestone 2 - target [undefined] - ESM in CJS

**Goal: Make ESM work in CJS**

- [ ] dynamically load ESM with a loader (ex @std/esm)
- [ ] transpile ESM to CJS (ex babel, webpack, etc.)
- [ ] ?

**Details**

There are a lot of unknowns here. Especially, until *any* first-class support is made available in Node.js.

Explore how this is being handled by the ecosystem:

- What workarounds are devs currently using in their projects?
- Do devs want to dynamically load ESM into CJS?
- Do devs want to provide alternative packages for CJS/ESM?
- Does this functionally really need to land into Node core?

Take into consideration that the front-end ecosystem has been transpiling code to be backward compatible for **years**. ES2015 -> ES5 transpilation is at least order of magnitude more complex to accomplish than ESM -> CJS. Some library devs may skip dynamic loading altogether and opt to provide alternative dist targets instead.

**Extensions**

- What are the pains of dynamic loading?
- Can hooks be used to make the loader path easier to implement?

### Milestone 3 - target [undefined] - CJS in ESM

**Goal: Make CJS work in ESM**

- [ ] add `import.meta` for ESM-compatible CJS built-ins

**Details:**

What needs to take place to make CJS mimic ESM in an ESM loaded module?

- [ ] can import.meta.require be made async
- [ ] does making it async require top-level await

CJS is the better 'known quantity' in the Node ecosystem. That's great! Now, what does it require to make `require()` work like ESM.

### Extension A - target [undefined] - Conversion paths for CJS -> ESM

**Goal: What does it take to convert CJS to ESM**

- [ ] document differences
- [ ] can this be automated?

**Details**

Exploration/research into what's required to make the conversion. May potentially lead to the 
development of tools or identifying/promoting tools provided by the greater ecosystem.

### Extension B - target [undefined] - Loader Hooks

**Goal: Address how loader hooks can be implemented/used within the current scope**

### Extension C - target [undefined] - Loader Extensions

**Goal: Addressing interop with existing loader extension workflows (ex APM)**

### Extension D - target [undefined] - `npm asset` interop

`npm asset` would be used on the front-end only. The browser has already implemented functionality to target and load ESM via relative paths. Does this even need to be mentioned here?

**Comment by mcollina** [comment]
I think loading a CJS module from ESM is something critical for 10.0.

**Comment by MylesBorins** [comment]
@evanplaice could you open a separate issue regarding roadmap? I would like to keep this issue scoped to... scope 😊

**Comment by ljharb** [comment]
I also think that "It does not include ESM<->CJS interop" is unacceptable for any unflagged implementation.

**Comment by evanplaice** [comment]
@MylesBorins Yes, I'll delete the above comment and add it as a separate issue.

**Comment by MylesBorins** [comment]
going to close this for now. We can revisit this (in a fresh thread) when we have a better idea of the exact work we will be doing
